In [1]:
import numpy as np
import pandas as pd
import os
import dotenv
import sqlalchemy

In [2]:
dotenv.load_dotenv(dotenv_path="cred.env", override=True)

username = os.getenv("username")
password = os.getenv("password")
host = os.getenv("host")
dbname = os.getenv("dbname")

connection_string = f"mysql+pymysql://{username}:{password}@{host}/{dbname}"
print(connection_string)

db_engine = sqlalchemy.create_engine(connection_string)

mysql+pymysql://studente_dapt:Ep1c0d3!!D4t4**4n4lys1s@epicode-data-pt-mysql.cvetyjye2qbl.eu-central-1.rds.amazonaws.com/AdventureWorksDW


In [3]:
query = "SELECT EnglishProductName, DealerPrice FROM dimproduct"
dimproduct = pd.read_sql(query, db_engine)

In [4]:
print(dimproduct["DealerPrice"].isna().sum(),"Nulli per colonna DealerPrice")
print(dimproduct.shape[0],"righe")
(dimproduct["DealerPrice"].isna().sum())*100/(dimproduct.shape[0])

211 Nulli per colonna DealerPrice
606 righe


34.81848184818482

*Pensavo de iniziare l'esercizio cancellando quello che sarebbero i NaN ma sono quasi 35% dei valori in colonna, quindi non posso gestirli cosi.*

In [5]:
#Sulla colonna DealerPrice, utilizzando il metodo .round(), arrotondiamo i valori alle due cifre decimali,
#e poi al valore intero più vicino
dimproduct["DealerPrice"] = dimproduct["DealerPrice"].round(2)
dimproduct.sample(5)

EnglishProductName  DealerPrice
14                        Decal 2          NaN
32       Front Derailleur Linkage          NaN
433  ML Road Frame-W - Yellow, 44       356.90
12                    Chain Stays          NaN
410       ML Mountain Front Wheel       125.42

In [6]:
dimproduct["DealerPrice"] = dimproduct["DealerPrice"].round(0)
dimproduct.sample(5)

EnglishProductName  DealerPrice
236     Long-Sleeve Logo Jersey, XL         30.0
162                   Metal Sheet 7          NaN
274         ML Road Frame - Red, 48        357.0
219          Sport-100 Helmet, Blue         20.0
308  HL Mountain Frame - Silver, 38        819.0

In [7]:
dimproduct["DealerPrice"] = dimproduct["DealerPrice"].clip(lower=0, upper=1000)
print(dimproduct["DealerPrice"].max(), "massimo presente")
print(dimproduct["DealerPrice"].min(),"minimo presente")

1000.0 massimo presente
1.0 minimo presente


In [8]:
#Calcola la somma cumulativa dei guadagni utilizzando il metodo .cumsum()
years = 5 
guadagni = pd.DataFrame({"Mese": list("GFMAMGLASOND"*years), 
                                   "Anno": np.repeat(list(range(years)), 12), 
                                   "Valore": np.random.randint(800, 5000, 12*years)})

guadagni["Cumulativo"] = guadagni["Valore"].cumsum()

#Come sopra, ma raggruppato per ogni anno usando prima un .groupby()
guadagni["Annuali"] = guadagni.groupby("Anno")["Valore"].cumsum()
print(guadagni)

   Mese  Anno  Valore  Cumulativo  Annuali
0     G     0    3998        3998     3998
1     F     0    1859        5857     5857
2     M     0    3155        9012     9012
3     A     0    1693       10705    10705
4     M     0    1376       12081    12081
5     G     0    2705       14786    14786
6     L     0    3189       17975    17975
7     A     0    1314       19289    19289
8     S     0    1173       20462    20462
9     O     0    4163       24625    24625
10    N     0    4907       29532    29532
11    D     0    1917       31449    31449
12    G     1    1456       32905     1456
13    F     1    3618       36523     5074
14    M     1     921       37444     5995
15    A     1    2111       39555     8106
16    M     1    4816       44371    12922
17    G     1    4589       48960    17511
18    L     1    3592       52552    21103
19    A     1    2318       54870    23421
20    S     1    1111       55981    24532
21    O     1    3934       59915    28466
22    N    

In [9]:
query1 = "SELECT FirstName, LastName, EmailAddress, Phone, AddressLine1 FROM dimcustomer"
dimcostumer = pd.read_sql(query1,db_engine)

#Trasformiamo i nomi dei clienti in modo che abbiano solo lettere minuscole,
#e i cognomi in modo che abbiano solo lettere maiuscole

dimcostumer["FirstName"] = dimcostumer["FirstName"].str.lower()
dimcostumer["LastName"] = dimcostumer["LastName"].str.upper()


In [10]:
#Sulla colonna EmailAddress, utilizzando il metodo .str.split(), estraiamo nome utente e dominio
dimcostumer[["User", "Domain"]] = dimcostumer["EmailAddress"].str.split('@', expand=True)
dimcostumer.sample(5)

FirstName  LastName                  EmailAddress                Phone  \
15500     billy      RUIZ    billy3@adventure-works.com         101-555-0159   
8818    brianna  ANDERSON  brianna8@adventure-works.com         394-555-0134   
9550    mohamed       PAL  mohamed1@adventure-works.com  1 (11) 500 555-0174   
2031     hunter      SHAN  hunter26@adventure-works.com         653-555-0115   
7014      wayne        XU    wayne6@adventure-works.com  1 (11) 500 555-0160   

               AddressLine1      User               Domain  
15500        Rt. 6415 Box A    billy3  adventure-works.com  
8818     917 Mark Twain Dr.  brianna8  adventure-works.com  
9550   3633 Stratton Circle  mohamed1  adventure-works.com  
2031     6671 Santa Barbara  hunter26  adventure-works.com  
7014   2822, rue des Ecoles    wayne6  adventure-works.com

In [11]:
#Sulla colonna Phone, estraiamo ogni parte del numero
#(ad es. da "1 (11) 500 555-0162" a ["1", "(11)", "500", "555-0162"])
dimcostumer[["CountryCode", "AreaCode", "Prefix", "LineNumber"]] = dimcostumer["Phone"].str.extract(r"(\d+)\s+(\(\d+\))\s+(\d+)\s+([\d-]+)")
dimcostumer.sample(5)

FirstName   LastName                  EmailAddress                Phone  \
14611    deanna  RODRIGUEZ  deanna23@adventure-works.com  1 (11) 500 555-0136   
3851      karla       TANG    karla4@adventure-works.com  1 (11) 500 555-0152   
5206      cindy      LEWIS   cindy21@adventure-works.com  1 (11) 500 555-0195   
470         jay     SUAREZ     jay49@adventure-works.com  1 (11) 500 555-0195   
14587    isaiah      YOUNG  isaiah42@adventure-works.com  1 (11) 500 555-0115   

                 AddressLine1      User               Domain CountryCode  \
14611            9736 Montana  deanna23  adventure-works.com           1   
3851    611, rue de Longchamp    karla4  adventure-works.com           1   
5206             Holzstr 1333   cindy21  adventure-works.com           1   
470    Heidestieg Straße 8664     jay49  adventure-works.com           1   
14587             Roßstr 9928  isaiah42  adventure-works.com           1   

      AreaCode Prefix LineNumber  
14611     (11)    500   555-0136  
3851      (11)    500   555-0152  
5206      (11)    500   555-0195  
470       (11)    500   555-0195  
14587     (11)    500   555-0115

In [12]:
#Utilizzando il metodo .str.contains(), estraiamo tutti gli indirizzi e-mail che contengono il numero "21"
dimcostumer[dimcostumer["User"].str.contains("21")]

FirstName   LastName                    EmailAddress  \
64        chase       REED     chase21@adventure-works.com   
118       alvin       ZENG     alvin21@adventure-works.com   
130    caroline    RUSSELL  caroline21@adventure-works.com   
163     gabriel       WANG   gabriel21@adventure-works.com   
203        luis       DIAZ      luis21@adventure-works.com   
...         ...        ...                             ...   
18337    hector      GOMEZ    hector21@adventure-works.com   
18365    cheryl    CARLSON    cheryl21@adventure-works.com   
18435   stanley  RODRIGUEZ   stanley21@adventure-works.com   
18440    rafael        CAI    rafael21@adventure-works.com   
18480      nina       RAJI      nina21@adventure-works.com   

                     Phone           AddressLine1        User  \
64            892-555-0184     2721 Alexander Pl.     chase21   
118    1 (11) 500 555-0174    3853 Wildcat Circle     alvin21   
130           424-555-0137       3884 Bates Court  caroline21   
163           617-555-0150       8002 Crane Court   gabriel21   
203           171-555-0126  3682 Diablo View Road      luis21   
...                    ...                    ...         ...   
18337  1 (11) 500 555-0116       3788 Linden Lane    hector21   
18365  1 (11) 500 555-0161      Helsenbergbogen 6    cheryl21   
18435  1 (11) 500 555-0166        Zeiter Weg 2264   stanley21   
18440  1 (11) 500 555-0129           Hellweg 4754    rafael21   
18480  1 (11) 500 555-0146      9 Katherine Drive      nina21   

                    Domain CountryCode AreaCode Prefix LineNumber  
64     adventure-works.com         NaN      NaN    NaN        NaN  
118    adventure-works.com           1     (11)    500   555-0174  
130    adventure-works.com         NaN      NaN    NaN        NaN  
163    adventure-works.com         NaN      NaN    NaN        NaN  
203    adventure-works.com         NaN      NaN    NaN        NaN  
...                    ...         ...      ...    ...        ...  
18337  adventure-works.com           1     (11)    500   555-0116  
18365  adventure-works.com           1     (11)    500   555-0161  
18435  adventure-works.com           1     (11)    500   555-0166  
18440  adventure-works.com           1     (11)    500   555-0129  
18480  adventure-works.com           1     (11)    500   555-0146  

[431 rows x 11 columns]

In [13]:
#Estraiamo tutti gli indirizzi e-mail che contengono il numero "20" oppure il numero "10"
dimcostumer[(dimcostumer["User"].str.contains("20"))|(dimcostumer["User"].str.contains("10"))]

FirstName LastName                      EmailAddress  \
1          eugene    HUANG      eugene10@adventure-works.com   
10      jacquelyn   SUAREZ   jacquelyn20@adventure-works.com   
22          ethan    ZHANG       ethan20@adventure-works.com   
30        bethany     YUAN     bethany10@adventure-works.com   
32         denise    STONE      denise10@adventure-works.com   
...           ...      ...                               ...   
18395       lydia     RANA       lydia10@adventure-works.com   
18397      walter    RAMOS      walter10@adventure-works.com   
18415       oscar   FLORES       oscar20@adventure-works.com   
18418     lindsey   SHARMA     lindsey10@adventure-works.com   
18428  kristopher    PEREZ  kristopher20@adventure-works.com   

                     Phone                    AddressLine1          User  \
1      1 (11) 500 555-0110                      2243 W St.      eugene10   
10     1 (11) 500 555-0169             7800 Corrinne Court   jacquelyn20   
22            589-555-0185             1769 Nicholas Drive       ethan20   
30     1 (11) 500 555-0182               636 Vine Hill Way     bethany10   
32     1 (11) 500 555-0169              626 Bentley Street      denise10   
...                    ...                             ...           ...   
18395  1 (11) 500 555-0182     182, avenue des Laurentides       lydia10   
18397  1 (11) 500 555-0143              2408, rue Maillard      walter10   
18415  1 (11) 500 555-0112  816, avenue des Champs-Elysées       oscar20   
18418  1 (11) 500 555-0134                     Zollstr 680     lindsey10   
18428  1 (11) 500 555-0183           Altendorfer Straße 39  kristopher20   

                    Domain CountryCode AreaCode Prefix LineNumber  
1      adventure-works.com           1     (11)    500   555-0110  
10     adventure-works.com           1     (11)    500   555-0169  
22     adventure-works.com         NaN      NaN    NaN        NaN  
30     adventure-works.com           1     (11)    500   555-0182  
32     adventure-works.com           1     (11)    500   555-0169  
...                    ...         ...      ...    ...        ...  
18395  adventure-works.com           1     (11)    500   555-0182  
18397  adventure-works.com           1     (11)    500   555-0143  
18415  adventure-works.com           1     (11)    500   555-0112  
18418  adventure-works.com           1     (11)    500   555-0134  
18428  adventure-works.com           1     (11)    500   555-0183  

[1042 rows x 11 columns]

In [14]:
#Calcolare la lunghezza di ogni indirizzo e-mail ed estrarre i cinque più lunghi e i cinque più corti
dimcostumer["EmailLength"] = dimcostumer["EmailAddress"].str.len()
top5_lunghi = dimcostumer.sort_values("EmailLength", ascending=False).head(5)
top5_corti = dimcostumer.sort_values("EmailLength", ascending=True).head(5)
dimcostumer.sample(5)

FirstName   LastName                   EmailAddress  \
16782     mason  HERNANDEZ    mason41@adventure-works.com   
1986     rachel     GARCIA   rachel19@adventure-works.com   
9200       noah   GONZALES     noah14@adventure-works.com   
145     jasmine    COLEMAN  jasmine46@adventure-works.com   
4023       juan     WATSON     juan10@adventure-works.com   

                     Phone           AddressLine1       User  \
16782  1 (11) 500 555-0114  1292 Bola Raton Court    mason41   
1986          429-555-0199         8403 Esperanza   rachel19   
9200          142-555-0173      9580 Kalima Place     noah14   
145           857-555-0115    1961 Oak Grove Road  jasmine46   
4023          175-555-0163       6409 Queens Road     juan10   

                    Domain CountryCode AreaCode Prefix LineNumber  EmailLength  
16782  adventure-works.com           1     (11)    500   555-0114           27  
1986   adventure-works.com         NaN      NaN    NaN        NaN           28  
9200   adventure-works.com         NaN      NaN    NaN        NaN           26  
145    adventure-works.com         NaN      NaN    NaN        NaN           29  
4023   adventure-works.com         NaN      NaN    NaN        NaN           26

In [15]:
print("Email più lunghe:")
print(top5_lunghi[["EmailAddress", "EmailLength"]])
print("\nEmail più corte:")
print(top5_corti[["EmailAddress", "EmailLength"]])

Email più lunghe:
                            EmailAddress  EmailLength
9242   christopher11@adventure-works.com           33
2655   christopher23@adventure-works.com           33
14964  christopher26@adventure-works.com           33
3512   christopher27@adventure-works.com           33
17017  christopher15@adventure-works.com           33

Email più corte:
                   EmailAddress  EmailLength
3974     k0@adventure-works.com           22
15299    y0@adventure-works.com           22
1658   joy2@adventure-works.com           24
550    deb4@adventure-works.com           24
13888  roy9@adventure-works.com           24


In [16]:
#• Modificare il dominio degli indirizzi e-mail da "adventure-works.com" 
#a "aw-db.com" mediante il metodo .str.replace()

dimcostumer["EmailAddress"] = dimcostumer["EmailAddress"].str.replace("adventure-works.com","aw-db.com")
dimcostumer.sample(5)

FirstName LastName         EmailAddress                Phone  \
15237     grace   MARTIN    grace14@aw-db.com         876-555-0124   
12338     lucas  JENKINS    lucas56@aw-db.com         708-555-0164   
2087     jeremy  COLLINS   jeremy19@aw-db.com         779-555-0186   
6874    natasha  ALVAREZ   natasha5@aw-db.com  1 (11) 500 555-0118   
9675    natalie   MARTIN  natalie82@aw-db.com         764-555-0123   

                AddressLine1       User               Domain CountryCode  \
15237        2885 Hill Drive    grace14  adventure-works.com         NaN   
12338     2423 Brookview Dr.    lucas56  adventure-works.com         NaN   
2087      6296 Elmonte Drive   jeremy19  adventure-works.com         NaN   
6874   570, quai de Grenelle   natasha5  adventure-works.com           1   
9675       1788 Camino Verde  natalie82  adventure-works.com         NaN   

      AreaCode Prefix LineNumber  EmailLength  
15237      NaN    NaN        NaN           27  
12338      NaN    NaN        NaN           27  
2087       NaN    NaN        NaN           28  
6874      (11)    500   555-0118           28  
9675       NaN    NaN        NaN           29

In [17]:
#Dalla colonna AddressLine1 estraiamo tutti gli indirizzi che contengono la sottostringa "Street"
dimcostumer[dimcostumer["AddressLine1"].str.contains("Street")]

FirstName   LastName         EmailAddress                Phone  \
7         marco      MEHTA    marco14@aw-db.com  1 (11) 500 555-0126   
12       lauren     WALKER   lauren41@aw-db.com         717-555-0164   
32       denise      STONE   denise10@aw-db.com  1 (11) 500 555-0169   
63       angela     MURPHY   angela41@aw-db.com         451-555-0162   
125        dana    NAVARRO      dana2@aw-db.com  1 (11) 500 555-0172   
...         ...        ...                  ...                  ...   
18273   timothy      LOPEZ  timothy38@aw-db.com         137-555-0130   
18292     diana  DOMINGUEZ    diana12@aw-db.com  1 (11) 500 555-0117   
18323    andrea     COOPER   andrea17@aw-db.com         586-555-0120   
18333      abby   MARTINEZ     abby15@aw-db.com  1 (11) 500 555-0150   
18351     jason       HALL    jason49@aw-db.com         158-555-0156   

                 AddressLine1       User               Domain CountryCode  \
7            942 Brook Street    marco14  adventure-works.com           1   
12          4785 Scott Street   lauren41  adventure-works.com         NaN   
32         626 Bentley Street   denise10  adventure-works.com           1   
63         4927 Virgil Street   angela41  adventure-works.com         NaN   
125        8481 Zartop Street      dana2  adventure-works.com           1   
...                       ...        ...                  ...         ...   
18273      2155 Zartop Street  timothy38  adventure-works.com         NaN   
18292       3931 Scott Street    diana12  adventure-works.com           1   
18323  134 Peachwillow Street   andrea17  adventure-works.com         NaN   
18333     1023 Hawkins Street     abby15  adventure-works.com           1   
18351        8618 Rose Street    jason49  adventure-works.com         NaN   

      AreaCode Prefix LineNumber  EmailLength  
7         (11)    500   555-0126           27  
12         NaN    NaN        NaN           28  
32        (11)    500   555-0169           28  
63         NaN    NaN        NaN           28  
125       (11)    500   555-0172           25  
...        ...    ...        ...          ...  
18273      NaN    NaN        NaN           29  
18292     (11)    500   555-0117           27  
18323      NaN    NaN        NaN           28  
18333     (11)    500   555-0150           26  
18351      NaN    NaN        NaN           27  

[512 rows x 12 columns]

In [18]:
#Dataset FACEBOOK
#Con la funzione pd.to_datetime() convertiamo la colonna status_published in formato Timestamp
facebook = pd.read_csv("Datasets/facebook.csv")
facebook["status_published"] = pd.to_datetime(facebook["status_published"])

#Utilizzando gli attributi .dt.year , .dt.month , .dt.day , .dt.dayofweek , .dt.dayofyear,
#ottieniamo informazioni specifiche sulle date delle transazioni, come l'anno, il mese,
#il giorno della settimana, il giorno dell'anno, eccetera

facebook["Anno"] = facebook["status_published"].dt.year
facebook["Mese"] = facebook["status_published"].dt.month
facebook["Giorno"] = facebook["status_published"].dt.day
facebook["GiornoSettimana"] = facebook["status_published"].dt.dayofweek     # Lunedì = 0
facebook["GiornoAnno"] = facebook["status_published"].dt.dayofyear

In [19]:
#Estraiamo solo i post relativi al 2012
facebook[facebook["Anno"] == 2012]

status_id status_type    status_published  \
2046  246675545449582_303160673134402       photo 2012-12-23 03:11:00   
2047  246675545449582_302341609882975       photo 2012-12-21 06:02:00   
2048  246675545449582_299212316862571       photo 2012-12-13 11:36:00   
2049  246675545449582_290159194434550       photo 2012-11-18 09:50:00   
2050  246675545449582_290154177768385       photo 2012-11-18 09:16:00   
...                               ...         ...                 ...   
2630  246675545449582_246688625448274       photo 2012-07-15 03:54:00   
2631  246675545449582_246686478781822       photo 2012-07-15 03:42:00   
2632  246675545449582_246684432115360       photo 2012-07-15 03:32:00   
2633  246675545449582_246678538782616       photo 2012-07-15 02:58:00   
2634  246675545449582_246677465449390       photo 2012-07-15 02:51:00   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
2046              0             0           0          0          0         0   
2047              4             1           0          4          0         0   
2048              3             0           0          3          0         0   
2049              5             1           1          5          0         0   
2050              2             0           0          2          0         0   
...             ...           ...         ...        ...        ...       ...   
2630             19            17           3         19          0         0   
2631             12             3           0         12          0         0   
2632             14             1           0         14          0         0   
2633             14             7           0         14          0         0   
2634             15             3           0         15          0         0   

      num_hahas  num_sads  num_angrys  Anno  Mese  Giorno  GiornoSettimana  \
2046          0         0           0  2012    12      23                6   
2047          0         0           0  2012    12      21                4   
2048          0         0           0  2012    12      13                3   
2049          0         0           0  2012    11      18                6   
2050          0         0           0  2012    11      18                6   
...         ...       ...         ...   ...   ...     ...              ...   
2630          0         0           0  2012     7      15                6   
2631          0         0           0  2012     7      15                6   
2632          0         0           0  2012     7      15                6   
2633          0         0           0  2012     7      15                6   
2634          0         0           0  2012     7      15                6   

      GiornoAnno  
2046         358  
2047         356  
2048         348  
2049         323  
2050         323  
...          ...  
2630         197  
2631         197  
2632         197  
2633         197  
2634         197  

[589 rows x 17 columns]

In [20]:
#Estraiamo solo i post relativi a maggio 2018
facebook[(facebook["Anno"] == 2018)&(facebook["Mese"] == 5)]

status_id status_type    status_published  \
2690    134115277150304_264554297439734       photo 2018-05-31 12:17:00   
2691    134115277150304_264493694112461       photo 2018-05-31 08:37:00   
2692    134115277150304_264416247453539       photo 2018-05-31 03:12:00   
2693    134115277150304_264168277478336       photo 2018-05-30 08:46:00   
2694    134115277150304_264160757479088       photo 2018-05-30 08:24:00   
...                                 ...         ...                 ...   
6339  1050855161656896_1615666458509094        link 2018-05-04 01:27:00   
6340  1050855161656896_1614717551937318       photo 2018-05-03 01:15:00   
6341  1050855161656896_1614187798656960       photo 2018-05-02 12:17:00   
6342  1050855161656896_1613063202102753       video 2018-05-01 08:38:00   
6343  1050855161656896_1612778858797854       photo 2018-05-01 01:18:00   

      num_reactions  num_comments  num_shares  num_likes  num_loves  num_wows  \
2690             11             0           0         10          0         1   
2691             56             4           2         55          1         0   
2692             72            13           1         69          3         0   
2693             10             0           0         10          0         0   
2694             56             3          12         54          2         0   
...             ...           ...         ...        ...        ...       ...   
6339             12             7           7         12          0         0   
6340             63            21           1         62          1         0   
6341            218           105         153        212          2         4   
6342            246          2373         475        108        130         5   
6343             34             5           0         33          1         0   

      num_hahas  num_sads  num_angrys  Anno  Mese  Giorno  GiornoSettimana  \
2690          0         0           0  2018     5      31                3   
2691          0         0           0  2018     5      31                3   
2692          0         0           0  2018     5      31                3   
2693          0         0           0  2018     5      30                2   
2694          0         0           0  2018     5      30                2   
...         ...       ...         ...   ...   ...     ...              ...   
6339          0         0           0  2018     5       4                4   
6340          0         0           0  2018     5       3                3   
6341          0         0           0  2018     5       2                2   
6342          2         1           0  2018     5       1                1   
6343          0         0           0  2018     5       1                1   

      GiornoAnno  
2690         151  
2691         151  
2692         151  
2693         150  
2694         150  
...          ...  
6339         124  
6340         123  
6341         122  
6342         121  
6343         121  

[595 rows x 17 columns]

In [21]:
#Confrontiamo il numero di post pubblicati nei weekend rispetto al numero di post pubblicati nel resto della settimana

facebook["FineSettimana"] = facebook["GiornoSettimana"].isin([5, 6])

# Conta i post weekend e non-weekend
weekend_counts = facebook["FineSettimana"].value_counts()

# Mostra i risultati
print("Post nel weekend:", weekend_counts[True])
print("Post in settimana:", weekend_counts[False])

Post nel weekend: 2041
Post in settimana: 5009


In [22]:
#Troviamo il primo e ultimo post pubblicati in ogni anno
primo_post_per_anno = facebook.groupby("Anno")["status_published"].min()
ultimo_post_per_anno = facebook.groupby("Anno")["status_published"].max()
print(primo_post_per_anno)
print("\n",ultimo_post_per_anno)

Anno
2012   2012-07-15 02:51:00
2013   2013-01-02 02:23:00
2014   2014-01-05 09:23:00
2015   2015-01-02 03:41:00
2016   2016-01-03 04:23:00
2017   2017-01-02 08:25:00
2018   2018-01-01 01:39:00
Name: status_published, dtype: datetime64[ns]

 Anno
2012   2012-12-23 03:11:00
2013   2013-12-31 23:38:00
2014   2014-12-10 07:42:00
2015   2015-12-31 08:20:00
2016   2016-12-31 05:13:00
2017   2017-12-31 22:15:00
2018   2018-06-13 01:12:00
Name: status_published, dtype: datetime64[ns]


In [23]:
#Quanti tipi di post ci sono? E quanti per ogni tipo?
tipi_post = facebook["status_type"].nunique()
print("Tipi di post:", tipi_post)
conteggio_per_tipo = facebook["status_type"].value_counts()
print("\n",conteggio_per_tipo)

Tipi di post: 4

 status_type
photo     4288
video     2334
status     365
link        63
Name: count, dtype: int64


In [24]:
pokemon = pd.read_csv("Datasets/pokemon.csv")
#Tramite i metodi .isnull() e .sum() controlliamo se ci sono valori nulli nel dataset 
#e contiamo quanti valori nulli ci sono in ogni colonna
pokemon.isna().sum()

#               0
Name            0
Type 1          0
Type 2        386
Total           0
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

Ci sono valori nulli? - si in type 2 troviamo 386 valori nulli
Se sì, avrebbe senso cercare di riempirli? -A quanto mi sembra non e' un dato che puo cambiare molto, perche si tratta del secondo tipo del pokemon, se fosse una cosa come "HP" (punti di vita) oppure Attack come esempi, allora in questi casi credo che sarebbe indispensabile il valore.

In [25]:
#Eliminiamo le righe che contengono valori nulli
pokemon.dropna()

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
6      6              Charizard     Fire  Flying    534  78      84       78   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
6        109       85    100           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[414 rows x 13 columns]

In [26]:
automobile = pd.read_csv("Datasets/automobile.csv")
#Ci sono valori nulli? Dove? Quanti?
automobile.isna().sum()

symboling             0
normalized-losses    37
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  0
stroke                0
compression-ratio     0
horsepower            0
peak-rpm              0
city-mpg              0
highway-mpg           0
price                 0
dtype: int64

In [27]:
automobile.loc[automobile["num-of-doors"].isna()]

symboling  normalized-losses   make fuel-type aspiration num-of-doors  \
26          1              148.0  dodge       gas      turbo          NaN   
60          0                NaN  mazda    diesel        std          NaN   

   body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
26      sedan          fwd           front        93.7  ...           98   
60      sedan          fwd           front        98.8  ...          122   

    fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
26         mpfi  3.03    3.39               7.6        102      5500       24   
60          idi  3.39    3.39              22.7         64      4650       36   

   highway-mpg  price  
26          30   8558  
60          42  10795  

[2 rows x 26 columns]

Parliamo di un Dataset di automobili dove le varianti di porte sono o 2 oppure 4, e guardando il tipo (body-style) rientrano soltanto in una di queste 2 categorie

In [28]:
print(automobile.shape)
filtro = automobile["body-style"] == "sedan"
automobile.loc[filtro,"num-of-doors"].value_counts()

(202, 26)


num-of-doors
four    79
two     14
Name: count, dtype: int64

questo caso si vede che per la variante 'sedan' la maggior parte delle auto sono tipo 'four' quindi tranquilamente userei questa categoria anche perche i valori cambiati sono solo 1% del dataset

In [29]:
# Il sensore a volte non funziona, dunque alcuni dati sono mancanti: 
#quale sarebbe la miglior strategia per gestirli?

temp = pd.DataFrame({"Giorno": [0, 1, 2, 3, 4, 5, 6 ,7, 8, 9, 10, 11, 12], 
                     "Temperature": [18, 19, 18, np.nan, 21, 20, 20, np.nan, 21, 23, np.nan, 23, 24]})

In questo caso credo che andrebbe benissimo fare interpolazione dei valori mancanti

In [30]:
temp["interp"] = temp["Temperature"].interpolate()
temp

Giorno  Temperature  interp
0        0         18.0    18.0
1        1         19.0    19.0
2        2         18.0    18.0
3        3          NaN    19.5
4        4         21.0    21.0
5        5         20.0    20.0
6        6         20.0    20.0
7        7          NaN    20.5
8        8         21.0    21.0
9        9         23.0    23.0
10      10          NaN    23.0
11      11         23.0    23.0
12      12         24.0    24.0

In [32]:
# Directory che contiene i dataset
directory = "Datasets/"

# Lista per salvare i nomi dei file che contengono dati nulli
file_con_nulli = []

# 1. Ciclo su tutti i file nella directory
for nome_file in os.listdir(directory):
    # 2. Controlla se il file ha estensione .csv
    if nome_file.endswith(".csv"):
        # 3. Leggi il file come DataFrame
        path_file = os.path.join(directory, nome_file)
        df = pd.read_csv(path_file)
        
        # 4. e 5. Calcola il numero totale di valori nulli
        numero_nulli = df.isna().sum().sum()
        
        # 6. Se ci sono dati nulli, salva il nome del file
        if numero_nulli > 0:
            file_con_nulli.append(nome_file)

# Stampa i file che contengono dati nulli
print("File con dati nulli:")
for f in file_con_nulli:
    print(f)


File con dati nulli:
automobile.csv
pokemon.csv
